File: backend/lambda/memory.py
"""
Lambda: POST /memory/save  — 「覚えて」トリガーでDynamoDBに保存
        GET  /memory/get   — 直近の記憶を取得して会話に使う
"""

import json
import os
import time
import uuid
import boto3
from boto3.dynamodb.conditions import Key

TABLE_NAME = os.environ.get("DYNAMODB_TABLE", "myagent-memory")
dynamodb   = boto3.resource("dynamodb")
table      = dynamodb.Table(TABLE_NAME)

CORS = {
    "Access-Control-Allow-Origin":  "*",
    "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
    "Access-Control-Allow-Headers": "Content-Type",
}


def lambda_handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return {"statusCode": 204, "headers": CORS, "body": ""}

    path   = event.get("path", "")
    method = event.get("httpMethod", "GET")

    if method == "POST" and "/save" in path:
        return _save(event)
    elif method == "GET" and "/get" in path:
        return _get(event)

    return _error(404, "Not Found")


def _save(event):
    try:
        body    = json.loads(event.get("body") or "{}")
        context_text = body.get("context", "").strip()
        keyword      = body.get("keyword", "").strip()
        user_id      = body.get("user_id", "default")
    except Exception:
        return _error(400, "Invalid JSON body")

    if not context_text:
        return _error(400, '"context" is required')

    item = {
        "user_id":   user_id,
        "timestamp": str(int(time.time() * 1000)),
        "id":        str(uuid.uuid4()),
        "keyword":   keyword,
        "context":   context_text,
    }

    try:
        table.put_item(Item=item)
    except Exception as e:
        return _error(500, f"DynamoDB保存エラー: {e}")

    return {
        "statusCode": 200,
        "headers":    {**CORS, "Content-Type": "application/json"},
        "body":       json.dumps({"saved": True}, ensure_ascii=False),
    }


def _get(event):
    user_id = event.get("queryStringParameters", {}).get("user_id", "default")
    limit   = int(event.get("queryStringParameters", {}).get("limit", "5"))

    try:
        result = table.query(
            KeyConditionExpression=Key("user_id").eq(user_id),
            ScanIndexForward=False,  # 新しい順
            Limit=limit,
        )
        items = result.get("Items", [])
    except Exception as e:
        return _error(500, f"DynamoDB取得エラー: {e}")

    return {
        "statusCode": 200,
        "headers":    {**CORS, "Content-Type": "application/json"},
        "body":       json.dumps({"memories": items}, ensure_ascii=False),
    }


def _error(status, message):
    return {
        "statusCode": status,
        "headers":    {**CORS, "Content-Type": "application/json"},
        "body":       json.dumps({"error": message}, ensure_ascii=False),
    }

File: functions/api/memory.js
/**
 * Cloudflare Pages Function — /api/memory
 * 「覚えて」で確定事実を保存／読み出し（会話の (A) 確定記憶）。
 *   GET    — 最新の記憶一覧（KV）
 *   POST   — 記憶を保存（KV ＋ 任意で D1 `memories` テーブル）
 *   DELETE — 全削除
 * バインディング: env.MEMORY = KV namespace。env.DB = D1（任意・あれば併用保存）。
 */
const CORS = {
  'Access-Control-Allow-Origin':  '*',
  'Access-Control-Allow-Methods': 'GET, POST, DELETE, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type',
};
const okRes  = d     => new Response(JSON.stringify(d),            { status: 200, headers: { ...CORS, 'Content-Type': 'application/json' } });
const errRes = (s, m) => new Response(JSON.stringify({ error: m }), { status: s,   headers: { ...CORS, 'Content-Type': 'application/json' } });

const KV_INDEX_KEY = 'memory_index';
const MAX_MEMORIES = 20;

export async function onRequest({ request, env }) {
  if (request.method === 'OPTIONS') return new Response(null, { status: 204, headers: CORS });

  const kv = env.MEMORY;
  if (!kv) return errRes(503, 'KV(MEMORY) が未設定です');

  if (request.method === 'GET') {
    const index    = JSON.parse((await kv.get(KV_INDEX_KEY)) || '[]');
    const memories = [];
    for (const id of index.slice(0, MAX_MEMORIES)) {
      const val = await kv.get(`memory:${id}`);
      if (val) memories.push(JSON.parse(val));
    }
    return okRes({ memories });
  }

  if (request.method === 'POST') {
    let body;
    try { body = await request.json(); } catch { return errRes(400, 'Invalid JSON'); }
    const keyword = (body?.keyword ?? '').toString().trim();
    if (!keyword) return errRes(400, '"keyword" is required');

    const id      = `${Date.now()}_${Math.random().toString(36).slice(2, 7)}`;
    const savedAt = new Date().toISOString();
    const item    = { id, keyword, context: (body?.context || keyword), saved_at: savedAt };

    // KV に保存（index 先頭に追加、上限超過は切り詰め）
    await kv.put(`memory:${id}`, JSON.stringify(item));
    const index = JSON.parse((await kv.get(KV_INDEX_KEY)) || '[]');
    index.unshift(id);
    if (index.length > MAX_MEMORIES) index.splice(MAX_MEMORIES);
    await kv.put(KV_INDEX_KEY, JSON.stringify(index));

    // D1 があれば併用保存（任意）
    if (env.DB) {
      try {
        await env.DB.prepare(
          'INSERT OR REPLACE INTO memories (id, keyword, content, saved_at) VALUES (?, ?, ?, ?)'
        ).bind(id, keyword, item.context, savedAt).run();
      } catch { /* memories テーブルが無くても続行 */ }
    }
    return okRes({ saved: true, id });
  }

  if (request.method === 'DELETE') {
    const index = JSON.parse((await kv.get(KV_INDEX_KEY)) || '[]');
    for (const id of index) await kv.delete(`memory:${id}`);
    await kv.delete(KV_INDEX_KEY);
    if (env.DB) { try { await env.DB.prepare('DELETE FROM memories').run(); } catch { /* ignore */ } }
    return okRes({ deleted: index.length });
  }

  return errRes(405, 'Method Not Allowed');
}
